### PR TITLE
DashboardScene: Alert states data layer

### LINKED ISSUE #77857: Alert state is not shown in the panel header

### PR DESCRIPTION
**Pull request by dprokop**
Closes https://github.com/grafana/grafana/issues/77857

This PR introduces support for alert states in DashboarScene. It implements the alerting workers (unified and legacy) as a single data layer.

For this PR to work, the complementary scenes PR is required: https://github.com/grafana/scenes/pull/454

TODO:
- [x]  Tests

### CODE REVIEW

File: public/app/features/dashboard-scene/scene/AlertStatesDataLayer.ts [hunk 1]
```diff
@@ -0,0 +1,227 @@
...
+          throw new Error(`Unexpected alert rules response.`);
+        })
+      );
+    }
+    this.querySub = alerStatesExecution.subscribe({
+      next: (stateUpdate) => {
+        this.publishResults(
+          {
+            state: LoadingState.Done,
+            series: [toDataFrame(stateUpdate)],
```
Comment (dprokop) [inline public/app/features/dashboard-scene/scene/AlertStatesDataLayer.ts:141]: Returned alert states are converted to data frame and handled accordingly in the SceneQueryRunner (filtered and rediced to a single alert state).

Comment (torkelo) [reply public/app/features/dashboard-scene/scene/AlertStatesDataLayer.ts:141]: @dprokop I have never seen toDataFrame called with a normal object array like this, what happens?

Comment (dprokop) [reply public/app/features/dashboard-scene/scene/AlertStatesDataLayer.ts:141]: Object keys will be translated to fields basically and each array value to a row. Here's the implementation: https://github.com/grafana/grafana/blob/dashboard-scene-alert-states/packages/grafana-data/src/dataframe/ArrayDataFrame.ts#L28

Comment (torkelo) [reply public/app/features/dashboard-scene/scene/AlertStatesDataLayer.ts:141]: ah, interesting. thought that required an explicit constructor call.

**Review by torkelo — APPROVED** [review]
Looks good, yes, some tests where it makes sense would be good

File: public/app/features/dashboard-scene/scene/AlertStatesDataLayer.ts [hunk 2]
```diff
@@ -0,0 +1,227 @@
...
+                  const state = promAlertStateToAlertState(rule.state);
+
+                  // there can be multiple alerts per panel, so we make sure we get the most severe state:
+                  // alerting > pending > ok
+                  if (!panelIdToAlertState[panelId]) {
+                    panelIdToAlertState[panelId] = {
+                      state,
+                      id: Object.keys(panelIdToAlertState).length,
+                      panelId,
+                      dashboardId: id!,
```
Comment (axelavargas) [inline public/app/features/dashboard-scene/scene/AlertStatesDataLayer.ts:112]: 🤔 shouldn't we use `uid` instead?

File: public/app/features/dashboard-scene/scene/AlertStatesDataLayer.ts [hunk 3]
```diff
@@ -0,0 +1,227 @@
...
+        )
+      ).pipe(map((alertStates) => alertStates));
+    } else {
+      alerStatesExecution = from(
+        getBackendSrv().get(
+          '/api/prometheus/grafana/api/v1/rules',
+          {
+            dashboard_uid: uid!,
+          },
+          `dashboard-query-runner-unified-alert-states-${id}`
```
Comment (axelavargas) [inline public/app/features/dashboard-scene/scene/AlertStatesDataLayer.ts:90]: Can we use the dashboard `uid` here as well? 
```suggestion
          `dashboard-query-runner-unified-alert-states-${uid}`
```

**Review by axelavargas — COMMENTED** [review]
Great job, @dprokop, I took a look and besides some minor comments and could grasp the general implementation of alerts in scenes. 

I also gave it a try locally, and found a minor issue, but I think is not related to this PR but to the scenes implementation.

image

File: public/app/features/dashboard-scene/scene/AlertStatesDataLayer.ts (hunk 3 above)
Comment (dprokop) [reply public/app/features/dashboard-scene/scene/AlertStatesDataLayer.ts:90]: This code is just a copy of what's currently implemented in UnifiedAlertStatesWorker.

File: public/app/features/dashboard-scene/scene/AlertStatesDataLayer.ts (hunk 2 above)
Comment (dprokop) [reply public/app/features/dashboard-scene/scene/AlertStatesDataLayer.ts:112]: This code is just a copy of what's currently implemented in UnifiedAlertStatesWorker.

**Comment by dprokop** [comment]
> I also gave it a try locally, and found a minor issue, but I think is not related to this PR but to the scenes implementation.

@axelavargas  - please add this as a bug to fix in https://github.com/grafana/grafana/issues/73993

**Review by axelavargas — APPROVED** [review]

**Pull request merged**